source/fem-dem/void_fraction.cc

```diff
@@ -0,0 +1,1141 @@
+                    {
+                      if (void_fraction_parameters
+                            ->qcm_sphere_equal_cell_volume == true)
+                        {
+                          // Get the radius by the volume of sphere which is
+                          // equal to the volume of cell
+                          r_sphere = radius_sphere_volume_cell(
+                            active_periodic_neighbors[n]->measure());
+                        }
+                      else
+                        {
+                          // The radius is obtained from the volume of sphere
+                          // based on R_s = h_omega
+                          r_sphere = radius_h_omega(
+                            active_periodic_neighbors[n]->measure());
+                        }
+                    }
+
+                  // Loop over quadrature points
+                  for (unsigned int k = 0; k < n_q_points; ++k)
+                    {
+                      // Adjust the location of the particle in the cell to
+                      // account for the periodicity. If the position of the
+                      // periodic cell if greater than the position of the
+                      // current cell, the particle location needs a positive
+                      // correction, and vice versa
+                      const Point<dim> particle_location =
+                        (active_periodic_neighbors[n]
+                           ->center()[periodic_direction] >
+                         cell->center()[periodic_direction]) ?
+                          particle.get_location() + periodic_offset :
+                          particle.get_location() - periodic_offset;
+
+                      // Distance between particle and quadrature point
+                      double periodic_neighbor_distance =
+                        particle_location.distance(
+                          periodic_neighbor_quadrature_point_location[n][k]);
+
+                      // Particle completely in reference sphere
+                      if (periodic_neighbor_distance <= (r_sphere - r_particle))
+                        {
+                          particle_properties
+                            [DEM::PropertiesIndex::volumetric_contribution] +=
+                            M_PI *
+                            Utilities::fixed_power<dim>(
+                              particle_properties[DEM::PropertiesIndex::dp]) /
+                            (2.0 * dim);
+                        }
+
+                      // Particle completely outside reference
+                      // sphere. Do absolutely nothing.
+
+                      // Particle partially in reference sphere
+                      else if ((periodic_neighbor_distance >
+                                (r_sphere - r_particle)) &&
+                               (periodic_neighbor_distance <
+                                (r_sphere + r_particle)))
```

Comment:
For readability, I would put the "completely outside" comment at the end or at the beginning. If at the beginning, I would say something like "Only cases where the particle is at least partially inside the reference sphere are treated." The gap in the  middle is a bit weird.

Reply:
Will try to fix that